File: Src/MusicPlayerBackend/Controllers/TrackController.cs
using System.Net.Mime;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using MusicPlayerBackend.Data;
using MusicPlayerBackend.Data.Entities;
using MusicPlayerBackend.Data.Repositories;
using MusicPlayerBackend.Services;
using MusicPlayerBackend.TransferObjects;
using MusicPlayerBackend.TransferObjects.Track;
using TrackListItem = MusicPlayerBackend.TransferObjects.Track.TrackListItem;
using TrackVisibility = MusicPlayerBackend.Data.Entities.TrackVisibility;

namespace MusicPlayerBackend.Controllers;

[ApiController]
[Consumes(MediaTypeNames.Application.Json)]
[Produces(MediaTypeNames.Application.Json)]
[ProducesResponseType(StatusCodes.Status400BadRequest)]
[ProducesResponseType(typeof(UnauthorizedResponse), StatusCodes.Status401Unauthorized)]
[Authorize]
[Route("[controller]")]
public sealed class TrackController(
    ITrackRepository trackRepository,
    IS3Service s3Service,
    IUnitOfWork unitOfWork,
    IUserProvider userProvider) : ControllerBase
{
    /// <summary>
    /// Gets track list
    /// </summary>
    /// <param name="request"></param>
    /// <param name="ct"></param>
    /// <returns></returns>
    [HttpGet(Name = "GetTracks")]
    [ProducesResponseType(typeof(TrackListResponse), StatusCodes.Status200OK)]
    public async Task<IActionResult> List([FromQuery] PlaylistListRequest request, CancellationToken ct)
    {
        var userId = await userProvider.GetUserIdAsync();

        var tracks = await trackRepository.QueryAll()
            .Where(t => t.OwnerUserId == userId)
            .Skip(request.Page * request.PageSize)
            .Take(request.PageSize)
            .Select(t => new TrackListItem
            {
                Author = t.Author,
                CoverUri = t.CoverUri,
                Name = t.Name,
                OwnerUserId = t.OwnerUserId,
                TrackUri = t.TrackUri,
                Visibility = (MusicPlayerBackend.TransferObjects.Track.TrackVisibility)(int)t.Visibility
            })
            .ToArrayAsync(ct);

        var trackCount = await trackRepository.QueryAll()
            .Where(t => t.OwnerUserId == userId)
            .CountAsync(ct);

        return Ok(new TrackListResponse { Items = tracks, Count = trackCount });
    }

    /// <summary>
    ///     Get track by Id
    /// </summary>
    /// <param name="id"></param>
    /// <param name="ct"></param>
    /// <returns></returns>
    [HttpGet("{id:guid}", Name = "GetTrack")]
    public async Task<IActionResult> Get(Guid id, CancellationToken ct)
    {
        var userId = await userProvider.GetUserIdAsync();
        var track = await trackRepository.GetByIdIfVisibleOrDefault(id, userId, ct);

        if (track == default)
            return BadRequest();

        return Ok(track);
    }

    [HttpPut("{id:guid}", Name = "UpdateTrack")]
    public async Task<IActionResult> Update(Guid id, TrackUpdateRequest request, CancellationToken ct)
    {
        var userId = await userProvider.GetUserIdAsync();
        var track = await trackRepository.GetByIdIfOwnerOrDefault(userId, id, ct);

        if (track == default)
            return BadRequest();

        track.Visibility = (TrackVisibility)request.Visibility;
        track.CoverUri = request.CoverUri;

        trackRepository.Save(track);
        await unitOfWork.SaveChangesAsync(ct);

        return NoContent();
    }

    [HttpDelete("{id:guid}", Name = "DeleteTrack")]
    public async Task<IActionResult> Delete(Guid id, CancellationToken ct)
    {
        var userId = await userProvider.GetUserIdAsync();
        var track = await trackRepository.GetByIdIfOwnerOrDefault(id, userId, ct);

        if (track == default)
            return BadRequest();

        trackRepository.Delete(track);
        await unitOfWork.SaveChangesAsync(ct);

        return NoContent();
    }

    [Consumes(MediaTypeNames.Multipart.FormData)]
    [Produces(MediaTypeNames.Application.Json)]
    [HttpPost(Name = "CreateTrack")]
    public async Task<IActionResult> Create([FromForm] TrackCreateRequest request, CancellationToken ct = default)
    {
        var uploadedTrackUri = await s3Service.TryUploadFileStream("tracks", request.Name, request.Track.OpenReadStream(), Path.GetExtension(request.Track.FileName), ct);

        if (uploadedTrackUri == default)
            return BadRequest(new { Error = "Failed to upload track" });

        string? coverUri = default;
        if (request.Cover is not null)
        {
            var uploadedCoverUri = await s3Service.TryUploadFileStream("covers", request.Name, request.Cover.OpenReadStream(), Path.GetExtension(request.Cover.FileName), ct);
            if (uploadedCoverUri == default)
                return BadRequest("Failed to upload cover");

            coverUri = uploadedCoverUri;
        }

        var trackEntity = new Track
        {
            CoverUri = coverUri,
            Name = request.Name,
            Author = request.Author,
            Visibility = (TrackVisibility)request.Visibility,
            TrackUri = uploadedTrackUri,
            OwnerUserId = await userProvider.GetUserIdAsync(),
        };

        trackRepository.Save(trackEntity);
        await unitOfWork.SaveChangesAsync(ct);

        return Ok(trackEntity);
    }
}

File: Src/Host/Controllers/TrackController.fs
namespace MusicPlayerBackend.Host.Controllers

open System
open System.IO
open System.Net.Mime
open System.Threading
open Microsoft.AspNetCore.Authorization
open Microsoft.AspNetCore.Http
open Microsoft.AspNetCore.Mvc

open MusicPlayerBackend.Common
open MusicPlayerBackend.Host
open MusicPlayerBackend.Host.Models
open MusicPlayerBackend.Host.Models.Track
open MusicPlayerBackend.Host.Services
open MusicPlayerBackend.Persistence
open MusicPlayerBackend.Persistence.Entities

[<ApiController>]
[<Consumes(MediaTypeNames.Application.Json)>]
[<Produces(MediaTypeNames.Application.Json)>]
[<ProducesResponseType(StatusCodes.Status400BadRequest)>]
[<ProducesResponseType(typeof<UnauthorizedResponse>, StatusCodes.Status401Unauthorized)>]
[<Authorize>]
[<Route("[controller]")>]
type TrackController(trackRepository: FsharpTrackRepository,
                     s3Service: S3Service,
                     unitOfWork: FsharpUnitOfWork,
                     userProvider: UserProvider) =
    inherit ControllerBase()

    /// <summary>
    ///     Gets track by Id.
    /// </summary>
    [<HttpGet("{id:guid}", Name = "GetTrack")>]
    [<ProducesResponseType(typeof<TrackResponse>, StatusCodes.Status200OK)>]
    member this.Get(id: TrackId, ct: CancellationToken) = task {
        let! userId = userProvider.GetUserId()
        let! track = trackRepository.TryGetVisible(id, userId, ct)
        match track with
        | None ->
            return this.NotFound() :> IActionResult
        | Some track ->
            let visibility =
                match track.Visibility with
                | TrackVisibility.Hidden -> Hidden
                | TrackVisibility.Visible -> Public
                | _ -> ArgumentOutOfRangeException() |> raise

            return this.Ok({
                Id = track.Id
                CoverUri = track.CoverUri
                TrackUri = track.TrackUri
                Visibility = visibility
                Name = track.Name
                Author = track.Author
                CreatedAt = track.CreatedAt
                UpdatedAt = track.UpdatedAt
            } : TrackResponse) :> IActionResult
    }

    /// <summary>
    ///     Updates track's visibility level, cover.
    /// </summary>
    [<HttpPut("{id:guid}", Name = "UpdateTrack")>]
    [<Consumes(MediaTypeNames.Application.FormUrlEncoded)>]
    member this.Update(id: Guid, [<FromForm>] request: UpdateTrackRequest, ct: CancellationToken) = task {
        let! userId = userProvider.GetUserId()
        let! track = trackRepository.TryGetIfOwner(userId, id, ct)
        match track with
        | None ->
            return this.NotFound() :> IActionResult
        | Some track ->
            match request.Visibility with
            | Some visibility ->
                track.Visibility <-
                    match visibility with
                    | Hidden -> TrackVisibility.Hidden
                    | Public -> TrackVisibility.Visible
            | _ -> ()

            match request.Cover with
            | Some cover ->
                if request.RemoveCover then
                    return this.BadRequest({
                        Error = $"{nameof(request.RemoveCover)} is true"
                    } : BadResponse) :> IActionResult
                else
                    let! uploadedCoverUri =
                        s3Service
                            .TryUploadFileStream(
                                "covers",
                                Guid.NewGuid().ToString() + "_" + track.Name,
                                cover.OpenReadStream(),
                                Path.GetExtension(cover.FileName)
                            )

                    if Option.isNone uploadedCoverUri then
                        return this.BadRequest({
                            Error = "Failed to upload cover"
                        } : BadResponse) :> IActionResult
                    else
                        track.CoverUri <- uploadedCoverUri
                        if request.RemoveCover then
                            track.CoverUri <- None
                            %trackRepository.Save(track)
                            do! unitOfWork.SaveChanges(ct)

                        return this.NoContent() :> IActionResult
            | None ->
                return this.NoContent() :> IActionResult
    }

    /// <summary>
    ///     Deletes track.
    /// </summary>
    [<HttpDelete("{id:guid}", Name = "DeleteTrack")>]
    member this.Delete(id: Guid, ct: CancellationToken) = task {
        let! userId = userProvider.GetUserId()
        do! trackRepository.DeleteIfOwner(id, userId, ct)
        do! unitOfWork.SaveChanges(ct)
        return this.NoContent() :> IActionResult
    }

    /// <summary>
    ///     Uploads track.
    /// </summary>
    [<HttpPost(Name = "UploadTrack")>]
    [<Consumes(MediaTypeNames.Multipart.FormData)>]
    [<ProducesResponseType(typeof<TrackResponse>, StatusCodes.Status200OK)>]
    member this.Upload([<FromForm>] request: CreateTrackRequest, ct: CancellationToken) = task {
        let! uploadedTrackUri = s3Service.TryUploadFileStream(
            "tracks",
            Guid.NewGuid().ToString() + "_" + request.Name,
            request.Track.OpenReadStream(),
            Path.GetExtension(request.Track.FileName))

        if Option.isNone uploadedTrackUri then
            return this.BadRequest({
                Error = "Failed to upload track"
            } : BadResponse) :> IActionResult
        else
            let mutable coverUri = null
            match request.Cover with
            | Some cover ->
                let! uploadedCoverUri = s3Service.TryUploadFileStream(
                    "covers", Guid.NewGuid().ToString() + "_" + request.Name,
                    cover.OpenReadStream(),
                    Path.GetExtension(cover.FileName))

                if Option.isNone uploadedCoverUri then
                    return this.BadRequest({
                        Error = "Failed to upload cover"
                    } : BadResponse) :> IActionResult
                else
                    coverUri <- uploadedCoverUri.Value

                    let! ownerUserId = userProvider.GetUserId()
                    let xv =
                        match request.Visibility with
                        | Hidden -> TrackVisibility.Hidden
                        | Public -> TrackVisibility.Visible

                    let track = Track.Create(
                        ownerUserId,
                        (coverUri |> Option.ofObj),
                        uploadedTrackUri.Value,
                        xv,
                        request.Name,
                        request.Author)

                    %trackRepository.Save(track)
                    do! unitOfWork.SaveChanges(ct)
                    return this.Ok({ Id = track.Id
                                     CoverUri = track.CoverUri
                                     TrackUri = track.TrackUri
                                     Author = track.Author
                                     Name = track.Name
                                     Visibility = request.Visibility
                                     CreatedAt = track.CreatedAt
                                     UpdatedAt = track.UpdatedAt } : TrackResponse) :> IActionResult
            | None ->
                return this.NoContent() :> IActionResult
    }


